server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CounterService.java

```diff
@@ -0,0 +1,77 @@
+package org.jboss.as.clustering.infinispan.subsystem;
+
+import org.infinispan.counter.EmbeddedCounterManagerFactory;
+import org.infinispan.counter.api.CounterConfiguration;
+import org.infinispan.counter.api.CounterManager;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.jboss.logging.Logger;
+import org.jboss.msc.service.Service;
+import org.jboss.msc.service.StartContext;
+import org.jboss.msc.service.StopContext;
+
+/**
+ * @author Vladimir Blagojevic
+ *
+ */
+public class CounterService implements Service<CounterConfiguration> {
+
+    private static final Logger log = Logger.getLogger(CounterService.class.getPackage().getName());
+
+    private final Dependencies dependencies;
+    private final String counterConfigurationName;
+    private final CounterConfiguration counterConfiguration;
+
+    public interface Dependencies {
+        EmbeddedCacheManager getCacheContainer();
+    }
+
+    public CounterService(CounterConfiguration configuration, String configurationName, Dependencies dependencies) {
+        this.counterConfiguration = configuration;
+        this.counterConfigurationName = configurationName;
+        this.dependencies = dependencies;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @see org.jboss.msc.value.Value#getValue()
+     */
+    @Override
+    public CounterConfiguration getValue() {
+        return this.counterConfiguration;
+    }
+
+    @Override
+    public void start(StartContext context) {
+        EmbeddedCacheManager container = this.dependencies.getCacheContainer();
+        CounterManager counterManager = EmbeddedCounterManagerFactory.asCounterManager(container);
+
+        //define counter
+        counterManager.defineCounter(counterConfigurationName, counterConfiguration);
+
+        //but also instantiate the counter as well
+        switch (counterConfiguration.type()) {
+            case BOUNDED_STRONG:
+            case UNBOUNDED_STRONG: {
+                counterManager.getStrongCounter(counterConfigurationName);
+                break;
+            }
+            case WEAK: {
+                counterManager.getWeakCounter(counterConfigurationName);
+                break;
+            }
+            default: {
+                log.warn("Unknown counter type " + counterConfiguration.type() + " did not get instantiated");
+                break;
+            }
+        }
+    }
+
+    @Override
+    public void stop(StopContext context) {
+        // remove and destroy the counter instance
+        EmbeddedCacheManager container = this.dependencies.getCacheContainer();
+        CounterManager counterManager = EmbeddedCounterManagerFactory.asCounterManager(container);
+        counterManager.remove(counterConfigurationName);
```

Comment:
`remove()` only removes the counter value. the counter is still defined. just put a comment to be aware in the future. an undefined operation is needed...

Reply:
I made a note

Reply:
this is the cause of the persistent problem. you are deleting the counter from the persistent storage!

Reply:
Ok @pruivo , so we remove only non-persistent? What is the right thing to do once stop is invoked?

Reply:
Ok, check this one @pruivo . You are right. Persistent counter,if not removed, preserves its value. So we simply do not remove counters from cache manager on stop?

Reply:
IMO, do nothing! if the counter is persistent, its value is stored in the file. if it is volatile, the reboot will clear its value from the memory.